README.md
# nv_themes

A theme changer plugin for [noveltree](https://peter88213.github.io/noveltree)


Adds a 'Theme Changer' entry to the 'Tools' menu to open a window with a combobox that lists all available themes. 

The selected theme will be persistently applied.  

For more information, see the [project homepage](https://peter88213.github.io/nv_themes) with description and download instructions.

## Feedback? Ideas? Feature requests?

You can go to the [noveltree "discussions" forum](https://github.com/peter88213/noveltree/discussions) and start a thread.

## Usage

Start the theme changer from the main menu: **Tools > Theme Changer**.

## License

This is Open Source software, and the *nv_themes* plugin is licensed under GPLv3. See the
[GNU General Public License website](https://www.gnu.org/licenses/gpl-3.0.en.html) for more
details, or consult the [LICENSE](https://github.com/peter88213/nv_themes/blob/main/LICENSE) file.


docs/changelog.md
[Project home page](index) > Changelog

------------------------------------------------------------------------

## Changelog

### v4.0.4

Fix a bug where the menu entry is disabled when no project is open.

Compatibility: novelyst v4.0 API

### v4.0.3

Refactor: replace tkinter constants with string literals.

Compatibility: novelyst v4.0 API

### v4.0.2

Fix a bug where the available themes are not displayed, if the *ttkthemes* package is not installed.

This is, because ttkthemes.ThemedStyle.theme_names() returns a list, whereas ttk.Style.theme_names() returns a tuple, which has no sort() method.

Compatibility: novelyst v4.0 API

### v4.0.1

- Make the setup script run with Python 3.11 under Windows.

Compatibility: novelyst v4.0 API

### v4.0.0

- API update. 

Compatibility: novelyst v4.0 API

### v3.0.0

- Code optimization and library update. 

Compatibility: novelyst v3.0 API

### v2.0.1

- API upgrade.

Compatibility: novelyst v2.0 API

### v0.2.1

- Release under the GPLv3 license.

Compatibility: novelyst v1.0 API


test/nv_themes.py
"""A 'Theme Changer' plugin for noveltree.

Version @release

Adds a 'Theme Changer' entry to the 'Tools' menu to open a window
with a combobox that lists all available themes. 
The selected theme will be persistently applied.  

To have a wider choice, you may want to install the ttkthemes package:

pip install ttkthemes

Requires Python 3.6+
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_themes
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
"""
import sys
import os
import tkinter as tk
from tkinter import ttk
import gettext
import locale
try:
    from ttkthemes import ThemedStyle
    extraThemes = True
except ModuleNotFoundError:
    extraThemes = False

LOCALE_PATH = f'{os.path.dirname(sys.argv[0])}/locale/'
try:
    CURRENT_LANGUAGE = locale.getlocale()[0][:2]
except:
    CURRENT_LANGUAGE = locale.getdefaultlocale()[0][:2]
try:
    t = gettext.translation('nv_themes', LOCALE_PATH, languages=[CURRENT_LANGUAGE])
    _ = t.gettext
except:

    def _(message):
        return message


class Plugin():
    VERSION = '@release'
    NOVELYST_API = '0.6'
    DESCRIPTION = 'Allows changing between available themes'
    URL = 'https://peter88213.github.io/nv_themes'

    def install(self, controller, ui, prefs):
        """Add a submenu to the 'Tools' menu.
        
        Positional arguments:
            controller -- reference to the main controller instance of the application.
            ui -- reference to the main view instance of the application.
        """
        self._ui = ui
        __, x, y = self._ui.root.geometry().split('+')
        offset = 300
        windowGeometry = f'+{int(x)+offset}+{int(y)+offset}'
        if extraThemes:
            self._ui.guiStyle = ThemedStyle(self._ui.root)
        if not self._ui.kwargs.get('gui_theme', ''):
            self._ui.kwargs['gui_theme'] = self._ui.guiStyle.theme_use()

        if not self._ui.kwargs['gui_theme'] in self._ui.guiStyle.theme_names():
            self._ui.kwargs['gui_theme'] = self._ui.guiStyle.theme_use()
        if extraThemes:
            self._ui.guiStyle.set_theme(self._ui.kwargs['gui_theme'])
        else:
            self._ui.guiStyle.theme_use(self._ui.kwargs['gui_theme'])

        self._ui.viewMenu.add_command(label=_('Change theme'), command=lambda: SettingsWindow(self._ui, windowGeometry))


class LabelCombo(ttk.Frame):
    """Combobox with a label.
    
    Credit goes to user stovfl on stackoverflow
    https://stackoverflow.com/questions/54584673/how-to-keep-tkinter-button-on-same-row-as-label-and-entry-box
    """

    def __init__(self, parent, text, textvariable, values, lblWidth=10):
        super().__init__(parent)
        self.pack(fill='x')
        self._label = ttk.Label(self, text=text, anchor='w', width=lblWidth)
        self._label.pack(side='left')
        self._combo = ttk.Combobox(self, textvariable=textvariable, values=values)
        self._combo.pack(side='left', fill='x', expand=True)

    def current(self):
        return self._combo.current()

    def configure(self, text=None, values=None):
        if text is not None:
            self._label['text'] = text
        if values is not None:
            self._combo['values'] = values


class SettingsWindow(tk.Toplevel):

    def __init__(self, ui, size, **kw):
        self._ui = ui
        super().__init__(**kw)
        self.title(_('Theme Changer'))
        self.geometry(size)
        self.grab_set()
        self.focus()
        window = ttk.Frame(self)
        window.pack(fill='both')

        theme = self._ui.guiStyle.theme_use()
        themeList = list(self._ui.guiStyle.theme_names())
        themeList.sort()
        self._theme = tk.StringVar(value=theme)
        self._theme.trace('w', self._change_theme)
        themeCombobox = LabelCombo(
            window,
            text=_('GUI Theme'),
            textvariable=self._theme,
            values=themeList,
            lblWidth=20
            )
        themeCombobox.pack(padx=5, pady=5)

        ttk.Button(window, text=_('Exit'), command=self.destroy).pack(padx=5, pady=5)

    def _change_theme(self, *args, **kwargs):
        theme = self._theme.get()
        self._ui.kwargs['gui_theme'] = theme
        if extraThemes:
            self._ui.guiStyle.set_theme(self._ui.kwargs['gui_theme'])
        else:
            self._ui.guiStyle.theme_use(self._ui.kwargs['gui_theme'])



tools/make_pot.py
"""Generate a template file (pot) for message translation.

For further information see https://github.com/peter88213/nv_themes
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
import build_nv_themes as build_plugin
sys.path.insert(0, f'{os.getcwd()}/../../novxlib/src')
import pgettext

APP = 'nv_themes'
POT_FILE = '../i18n/messages.pot'


def make_pot(version='unknown'):
    # Generate a complete plugin.
    build_plugin.main()

    # Generate a pot file from the script.
    if os.path.isfile(POT_FILE):
        os.replace(POT_FILE, f'{POT_FILE}.bak')
        backedUp = True
    else:
        backedUp = False
    try:
        pot = pgettext.PotFile(POT_FILE, app=APP, appVersion=version)
        pot.scan_file(build_plugin.TARGET_FILE)
        print(f'Writing "{pot.filePath}"...\n')
        pot.write_pot()

    except:
        if backedUp:
            os.replace(f'{POT_FILE}.bak', POT_FILE)
        print('WARNING: Cannot write pot file.')
        return False

    return True


if __name__ == '__main__':
    try:
        success = make_pot(sys.argv[1])
    except:
        success = make_pot()
    if not success:
        sys.exit(1)


tools/translate_de.py
"""Generate German translation files for GNU gettext.

- Update the project's 'de.po' translation file.
- Generate the language specific 'novxlib.mo' dictionary.

Usage: 
translate_de.py

File structure:

├── novxlib/
│   ├── i18n/
│   │   └── de.json
│   └── src/
│       ├── translations.py
│       └── msgfmt.py
└── nv_themes/
    ├── src/ 
    ├── tools/ 
    │   └── translate_de.py
    └── i18n/
        ├── messages.pot
        ├── de.po
        └── locale/
            └─ de/
               └─ LC_MESSAGES/
                  └─ nv_themes.mo
    
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_themes
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import translations
from shutil import copyfile
import msgfmt

APP_NAME = 'nv_themes'
PO_PATH = '../i18n/de.po'
MO_PATH = f'../i18n/locale/de/LC_MESSAGES/{APP_NAME}.mo'
MO_COPY = f'../../kalliope/src/locale/de/LC_MESSAGES/{APP_NAME}.mo'


def main(version='unknown'):
    if translations.main('de', app=APP_NAME, appVersion=version, json=True):
        print(f'Writing "{MO_PATH}" ...')
        msgfmt.make(PO_PATH, MO_PATH)
        copyfile(MO_PATH, MO_COPY)
    else:
        sys.exit(1)


if __name__ == '__main__':
    try:
        main(sys.argv[1])
    except:
        main()


src/novelyst_themes.py
"""A 'Theme Changer' plugin for novelyst.

Version @release

Adds a 'Theme Changer' entry to the 'Tools' menu to open a window
with a combobox that lists all available themes. 
The selected theme will be persistently applied.  

To have a wider choice, you may want to install the ttkthemes package:

pip install ttkthemes

Requires Python 3.6+
Copyright (c) 2022 Peter Triesberger
For further information see https://github.com/peter88213/novelyst_themes
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
"""
APPLICATION = 'Theme Changer'
PLUGIN = f'{APPLICATION} plugin @release'

import tkinter as tk
from tkinter import ttk
try:
    from ttkthemes import ThemedStyle
    extraThemes = True
except ModuleNotFoundError:
    extraThemes = False


class Plugin():
    """A 'Theme Changer' plugin class.
    
    Public methods:
        disable_menu() -- disable menu entries when no project is open.
        enable_menu() -- enable menu entries when a project is open.    
    """
    VERSION = '@release'
    NOVELYST_API = '4.0'
    DESCRIPTION = 'Allows changing between available themes'
    URL = 'https://peter88213.github.io/novelyst_themes'

    def install(self, ui):
        """Add a submenu to the 'Tools' menu.
        
        Positional arguments:
            ui -- reference to the NovelystTk instance of the application.
        """
        self._ui = ui
        __, x, y = self._ui.root.geometry().split('+')
        offset = 300
        windowGeometry = f'+{int(x)+offset}+{int(y)+offset}'
        if extraThemes:
            self._ui.guiStyle = ThemedStyle(self._ui.root)
        if not self._ui.kwargs.get('gui_theme', ''):
            self._ui.kwargs['gui_theme'] = self._ui.guiStyle.theme_use()

        if not self._ui.kwargs['gui_theme'] in self._ui.guiStyle.theme_names():
            self._ui.kwargs['gui_theme'] = self._ui.guiStyle.theme_use()
        if extraThemes:
            self._ui.guiStyle.set_theme(self._ui.kwargs['gui_theme'])
        else:
            self._ui.guiStyle.theme_use(self._ui.kwargs['gui_theme'])

        # Create a submenu
        self._ui.toolsMenu.add_command(label=APPLICATION, command=lambda: SettingsWindow(self._ui, windowGeometry))


class LabelCombo(ttk.Frame):
    """Combobox with a label.
    
    Credit goes to user stovfl on stackoverflow
    https://stackoverflow.com/questions/54584673/how-to-keep-tkinter-button-on-same-row-as-label-and-entry-box
    """

    def __init__(self, parent, text, textvariable, values, lblWidth=10):
        super().__init__(parent)
        self.pack(fill='x')
        self._label = ttk.Label(self, text=text, anchor='w', width=lblWidth)
        self._label.pack(side='left')
        self._combo = ttk.Combobox(self, textvariable=textvariable, values=values)
        self._combo.pack(side='left', fill='x', expand=True)

    def current(self):
        """Return the combobox selection."""
        return self._combo.current()

    def configure(self, text=None, values=None):
        """Configure internal widgets."""
        if text is not None:
            self._label['text'] = text
        if values is not None:
            self._combo['values'] = values


class SettingsWindow(tk.Toplevel):

    def __init__(self, ui, size, **kw):
        self._ui = ui
        super().__init__(**kw)
        self.title(APPLICATION)
        self.geometry(size)
        self.grab_set()
        self.focus()
        window = ttk.Frame(self)
        window.pack(fill='both')

        # Combobox for theme setting.
        theme = self._ui.guiStyle.theme_use()
        themeList = list(self._ui.guiStyle.theme_names())
        themeList.sort()
        self._theme = tk.StringVar(value=theme)
        self._theme.trace('w', self._change_theme)
        themeCombobox = LabelCombo(window,
                              text='GUI Theme',
                              textvariable=self._theme,
                              values=themeList,
                              lblWidth=20)
        themeCombobox.pack(padx=5, pady=5)

        # "Exit" button.
        ttk.Button(window, text='Exit', command=self.destroy).pack(padx=5, pady=5)

    def _change_theme(self, *args, **kwargs):
        theme = self._theme.get()
        self._ui.kwargs['gui_theme'] = theme
        if extraThemes:
            self._ui.guiStyle.set_theme(self._ui.kwargs['gui_theme'])
        else:
            self._ui.guiStyle.theme_use(self._ui.kwargs['gui_theme'])



tools/build_nv_themes.py
"""Build a nv_themes plugin.
        
In order to distribute a single script without dependencies, 
this script "inlines" all modules imported from the novxlib package.

The novxlib project (see https://github.com/peter88213/novxlib)
must be located on the same directory level as the nv_themes project. 

Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_themes
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import inliner

SRC = '../src/'
BUILD = '../test/'
SOURCE_FILE = f'{SRC}nv_themes.py'
TARGET_FILE = f'{BUILD}nv_themes.py'

os.makedirs(BUILD, exist_ok=True)


def main():
    inliner.run(SOURCE_FILE, TARGET_FILE, 'novxlib', '../../novxlib-Alpha/src/')
    print('Done.')


if __name__ == '__main__':
    main()


src/nv_themes.py
"""A 'Theme Changer' plugin for noveltree.

Version @release

Adds a 'Theme Changer' entry to the 'Tools' menu to open a window
with a combobox that lists all available themes. 
The selected theme will be persistently applied.  

To have a wider choice, you may want to install the ttkthemes package:

pip install ttkthemes

Requires Python 3.6+
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_themes
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
"""
import sys
import os
import tkinter as tk
from tkinter import ttk
import gettext
import locale
try:
    from ttkthemes import ThemedStyle
    extraThemes = True
except ModuleNotFoundError:
    extraThemes = False

# Initialize localization.
LOCALE_PATH = f'{os.path.dirname(sys.argv[0])}/locale/'
try:
    CURRENT_LANGUAGE = locale.getlocale()[0][:2]
except:
    # Fallback for old Windows versions.
    CURRENT_LANGUAGE = locale.getdefaultlocale()[0][:2]
try:
    t = gettext.translation('nv_themes', LOCALE_PATH, languages=[CURRENT_LANGUAGE])
    _ = t.gettext
except:

    def _(message):
        return message


class Plugin():
    """A 'Theme Changer' plugin class."""
    VERSION = '@release'
    NOVELYST_API = '0.7'
    DESCRIPTION = 'Allows changing between available themes'
    URL = 'https://peter88213.github.io/nv_themes'

    def install(self, model, view, controller, prefs):
        """Add a submenu to the 'Tools' menu.
        
        Positional arguments:
            controller -- reference to the main controller instance of the application.
            view -- reference to the main view instance of the application.
        """
        self._ui = view
        self._prefs = prefs
        __, x, y = self._ui.root.geometry().split('+')
        offset = 300
        windowGeometry = f'+{int(x)+offset}+{int(y)+offset}'
        if extraThemes:
            self._ui.guiStyle = ThemedStyle(self._ui.root)
        if not self._prefs.get('gui_theme', ''):
            self._prefs['gui_theme'] = self._ui.guiStyle.theme_use()

        if not self._prefs['gui_theme'] in self._ui.guiStyle.theme_names():
            self._prefs['gui_theme'] = self._ui.guiStyle.theme_use()
        if extraThemes:
            self._ui.guiStyle.set_theme(self._prefs['gui_theme'])
        else:
            self._ui.guiStyle.theme_use(self._prefs['gui_theme'])

        # Create a submenu
        self._ui.viewMenu.add_command(
            label=_('Change theme'),
            command=lambda: SettingsWindow(self._ui, self._prefs, windowGeometry)
            )


class LabelCombo(ttk.Frame):
    """Combobox with a label.
    
    Credit goes to user stovfl on stackoverflow
    https://stackoverflow.com/questions/54584673/how-to-keep-tkinter-button-on-same-row-as-label-and-entry-box
    """

    def __init__(self, parent, text, textvariable, values, lblWidth=10):
        super().__init__(parent)
        self.pack(fill='x')
        self._label = ttk.Label(self, text=text, anchor='w', width=lblWidth)
        self._label.pack(side='left')
        self._combo = ttk.Combobox(self, textvariable=textvariable, values=values)
        self._combo.pack(side='left', fill='x', expand=True)

    def current(self):
        """Return the combobox selection."""
        return self._combo.current()

    def configure(self, text=None, values=None):
        """Configure internal widgets."""
        if text is not None:
            self._label['text'] = text
        if values is not None:
            self._combo['values'] = values


class SettingsWindow(tk.Toplevel):

    def __init__(self, view, prefs, size, **kw):
        self._ui = view
        self._prefs = prefs
        super().__init__(**kw)
        self.title(_('Theme Changer'))
        self.geometry(size)
        self.grab_set()
        self.focus()
        window = ttk.Frame(self)
        window.pack(fill='both')

        # Combobox for theme setting.
        theme = self._ui.guiStyle.theme_use()
        themeList = list(self._ui.guiStyle.theme_names())
        themeList.sort()
        self._theme = tk.StringVar(value=theme)
        self._theme.trace('w', self._change_theme)
        themeCombobox = LabelCombo(
            window,
            text=_('GUI Theme'),
            textvariable=self._theme,
            values=themeList,
            lblWidth=20
            )
        themeCombobox.pack(padx=5, pady=5)

        # "Exit" button.
        ttk.Button(window, text=_('Exit'), command=self.destroy).pack(padx=5, pady=5)

    def _change_theme(self, *args, **kwargs):
        theme = self._theme.get()
        self._prefs['gui_theme'] = theme
        if extraThemes:
            self._ui.guiStyle.set_theme(self._prefs['gui_theme'])
        else:
            self._ui.guiStyle.theme_use(self._prefs['gui_theme'])

